pkg/frontend/v2/frontend_test.go

```diff
@@ -359,6 +359,56 @@ func TestFrontend_Protobuf_ResponseClosedBeforeStreamExhausted(t *testing.T) {
 	resp.Close() // We expect all goroutines to be cleaned up after this (verified by the VerifyNoLeakTestMain call in TestMain above)
 }
 
+func TestFrontend_Protobuf_ResponseClosedBeforeResponseReceived(t *testing.T) {
+	respChannel := make(chan *ProtobufResponseStream)
+	defer close(respChannel)
+
+	f, _ := setupFrontend(t, nil, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		if msg.Type != schedulerpb.ENQUEUE {
+			return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.ERROR, Error: fmt.Sprintf("unexpected message type %v sent to scheduler", msg.Type)}
+		}
+
+		go func() {
+			// Close the stream returned by DoProtobufRequest once we're confident the goroutine in DoProtobufRequest has observed that the request has been enqueued
+			// and is waiting for streamContext to be cancelled.
+			// This ensures that closing the channel doesn't trigger the code path that calls writeEnqueueError().
+			resp := <-respChannel
+			time.Sleep(10 * time.Millisecond)
```

Comment:
I wish we can do this without time.Sleep, but I believe that will make the code harder to read 🤔 . Hopefully this won't make the test flaky in CI.